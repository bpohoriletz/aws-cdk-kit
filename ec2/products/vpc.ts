import * as ec2 from 'aws-cdk-lib/aws-ec2';

export class VpcProduct implements IVpcProduct {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

interface IVpcProduct extends ec2.VpcProps {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

export interface IVpcBuilder {
  setCidr?(name: string): this;
  setAzs?(): this;
  setNatProvider?(): this;
  setNatSubnets?(): this;
  setSubnetConfiguration?(name: string): this;
  getResult(): ec2.VpcProps;
}

export interface IVpcBuilderConstructor {
  new (): IVpcBuilder;
}
