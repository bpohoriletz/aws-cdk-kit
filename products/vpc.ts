import * as ec2 from 'aws-cdk-lib/aws-ec2';

export class VpcProduct implements IVpcProduct {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

interface IVpcProduct extends ec2.VpcProps {
  ipAddresses?: ec2.IIpAddresses;
  maxAzs?: number;
  natGatewayProvider?: ec2.NatProvider;
  natGatewaySubnets?: ec2.SubnetSelection;
  subnetConfiguration?: ec2.SubnetConfiguration[];
}

export interface IVpcBuilder {
  customizeCidr?(name: string): this;
  customizeAzs?(): this;
  customizeNatProvider?(): this;
  customizeNatSubnets?(): this;
  customizeSubnetConfiguration?(name: string): this;
  getResult(): ec2.VpcProps;
}

export interface IVpcBuilderConstructor {
  new (): IVpcBuilder;
}
