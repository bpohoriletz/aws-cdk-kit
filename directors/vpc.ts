import { VpcProduct, IVpcBuilder, IVpcBuilderConstructor } from "../products/vpc";

export default class VpcDirector {
  private builder: IVpcBuilder;

  constructor(vpcName: string, builder: IVpcBuilderConstructor){
    this.builder = new builder(vpcName);
  }

  constructVpc(): VpcProduct {
    return this.builder
             .customizeAzs()
             .customizeCidr()
             .customizeNatProvider()
             .customizeNatSubnets()
             .customizeSubnetConfiguration()
             .getResult();
  }
}
