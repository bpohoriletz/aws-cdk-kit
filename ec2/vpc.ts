import * as cdk from "aws-cdk-lib";
import * as ec2 from "aws-cdk-lib/aws-ec2";
import * as names from "../naming/resources";
import * as conf from "../../config/ec2/vpc";

export function createMinimalVpc(resourceNamePrefix: string[], stack: cdk.Stack): ec2.Vpc {
  const resourceName = names.ec2VpcName(resourceNamePrefix);
  const vpc = new ec2.Vpc(stack, resourceName, {
    ipAddresses: conf.vpcCidrConfig(resourceName),
    maxAzs: 1,
    subnetConfiguration: conf.vpcSubnetConfig(resourceName),
    natGatewayProvider: ec2.NatProvider.instanceV2({
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T2,
        ec2.InstanceSize.NANO
      ),
    }),
    natGatewaySubnets: {
      subnetType: ec2.SubnetType.PUBLIC, // Make NAT instance reside in a public subnet
    },
  });

  return vpc;
}

export function createVpc(resourceNamePrefix: string[], stack: cdk.Stack): ec2.Vpc {
  const resourceName = names.ec2VpcName(resourceNamePrefix);
  const vpc = new ec2.Vpc(stack, resourceName, {
    ipAddresses: conf.vpcCidrConfig(resourceName),
    maxAzs: 2,             // Use 2 availability zones
    subnetConfiguration: conf.vpcSubnetConfig(resourceName),
    natGatewayProvider: ec2.NatProvider.instanceV2({
      instanceType: ec2.InstanceType.of(
        ec2.InstanceClass.T3,
        ec2.InstanceSize.MICRO
      ),
    }),
    natGatewaySubnets: {
      subnetType: ec2.SubnetType.PUBLIC, // Make NAT instance reside in a public subnet
    },
  });

  return vpc;
}

export function createProdVpc(resourceNamePrefix: string[], stack: cdk.Stack): ec2.Vpc {
  const resourceName = names.ec2VpcName(resourceNamePrefix);
  const vpc = new ec2.Vpc(stack, resourceName, {
    ipAddresses: conf.vpcCidrConfig(resourceName),
    subnetConfiguration: conf.vpcSubnetConfig(resourceName),
  });

  return vpc;
}
