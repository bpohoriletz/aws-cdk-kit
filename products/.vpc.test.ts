import { VpcProduct } from "./vpc";

describe("new VpcProduct", () => {
  let instance: VpcProduct;

  beforeEach(() => {
    instance = new VpcProduct();
  });

  test("has default properties set", () => {
    expect(instance).toMatchSnapshot();
  });
});

