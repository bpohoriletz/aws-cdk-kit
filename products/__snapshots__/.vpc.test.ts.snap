// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`new VpcProduct has default properties set 1`] = `VpcProduct {}`;
